File: packages/journal-encryption-lib/src/dek_cache.ts
import {
  DecryptCommand,
  GenerateDataKeyCommand,
  KMS,
  KMSClient,
} from "@aws-sdk/client-kms";
import { webcrypto } from "node:crypto";

interface EncryptingDek {
  key: webcrypto.CryptoKey;
  encryptedDek: Uint8Array;
}

export class DekCache {
  private kms: KMSClient;
  private encryptingKmsKeyID: string;
  private encryptingDek?: Promise<EncryptingDek>;
  private decryptingDeks: Map<string, Promise<webcrypto.CryptoKey>>;

  constructor({
    kms,
    encryptingKmsKeyID,
  }: {
    kms: KMSClient;
    encryptingKmsKeyID: string;
  }) {
    this.kms = kms;
    this.encryptingKmsKeyID = encryptingKmsKeyID;
    this.encryptingDek = undefined;
    this.decryptingDeks = new Map();
  }

  static cacheKey(encryptedDek: Uint8Array): string {
    return Buffer.from(encryptedDek).toString("base64");
  }

  cacheDecryptingDek(
    encryptedDek: Uint8Array,
    dek: Promise<webcrypto.CryptoKey>
  ) {
    const cacheKey = DekCache.cacheKey(encryptedDek);
    this.decryptingDeks.set(
      cacheKey,
      dek.catch((e) => {
        this.decryptingDeks.delete(cacheKey);
        return Promise.reject(e);
      })
    );
  }

  async getEncryptingDek(): Promise<EncryptingDek> {
    if (this.encryptingDek !== undefined) return this.encryptingDek;

    this.encryptingDek = this.createEncryptingDek()
      .then((dek) => {
        this.cacheDecryptingDek(
          dek.encryptedDek,
          Promise.resolve(dek.decryptingDek)
        );
        return { key: dek.encryptingDek, encryptedDek: dek.encryptedDek };
      })
      .catch((e) => {
        this.encryptingDek = undefined;
        return Promise.reject(e);
      });

    return this.encryptingDek;
  }

  async createEncryptingDek() {
    try {
      const start = new Date();
      const dekResult = await this.kms.send(
        new GenerateDataKeyCommand({
          KeyId: this.encryptingKmsKeyID,
          KeySpec: "AES_256", // 32 byte keys, 184 byte encrypted keys
        })
      );

      if (!dekResult.Plaintext) {
        throw new Error("Missing dek plaintext");
      }
      if (!dekResult.CiphertextBlob) {
        throw new Error("Missing dek ciphertext");
      }

      const [encryptingDek, decryptingDek] = await Promise.all([
        webcrypto.subtle.importKey(
          "raw",
          dekResult.Plaintext,
          "AES-GCM",
          false,
          ["encrypt"]
        ),
        webcrypto.subtle.importKey(
          "raw",
          dekResult.Plaintext,
          "AES-GCM",
          false,
          ["decrypt"]
        ),
      ]);

      console.log(
        `Created encrypting dek against KMS key '${
          this.encryptingKmsKeyID
        }' in ${new Date().valueOf() - start.valueOf()}ms`
      );

      return {
        encryptedDek: dekResult.CiphertextBlob,
        encryptingDek,
        decryptingDek,
      };
    } catch (e) {
      console.log(
        `Failed to create encrypting dek against KMS key '${this.encryptingKmsKeyID}':`,
        e
      );
      throw e;
    }
  }

  async getDecryptingDek(
    encryptedDek: Uint8Array
  ): Promise<webcrypto.CryptoKey> {
    const cacheKey = DekCache.cacheKey(encryptedDek);
    const cachedDek = this.decryptingDeks.get(cacheKey);
    if (cachedDek !== undefined) return cachedDek;

    const dekPromise = this.decryptDek(encryptedDek).catch((e) => {
      this.decryptingDeks.delete(cacheKey);
      return Promise.reject(e);
    });

    this.decryptingDeks.set(cacheKey, dekPromise);

    return dekPromise;
  }

  async decryptDek(encryptedDek: Uint8Array): Promise<webcrypto.CryptoKey> {
    try {
      const start = new Date();
      const dekResult = await this.kms.send(
        new DecryptCommand({
          CiphertextBlob: encryptedDek,
        })
      );

      if (!dekResult.Plaintext) {
        throw new Error("Missing dek plaintext");
      }

      const dek = await webcrypto.subtle.importKey(
        "raw",
        dekResult.Plaintext,
        "AES-GCM",
        false,
        ["decrypt"]
      );

      console.log(
        `Loaded decryping dek in ${new Date().valueOf() - start.valueOf()}ms`
      );

      return dek;
    } catch (e) {
      console.log("Failed to load decrypting dek:", e);
      throw e;
    }
  }
}

File: golang-sensitive/dek_cache.go
package sensitive

import (
	"context"
	"crypto/aes"
	"crypto/cipher"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/aws/aws-sdk-go-v2/service/kms"
	"github.com/aws/aws-sdk-go-v2/service/kms/types"
	lru "github.com/hashicorp/golang-lru/v2"
)

// EncryptingDek represents a data encryption key for encryption
type EncryptingDek struct {
	Key          cipher.AEAD
	EncryptedDek []byte
}

// DekCache manages data encryption keys with caching
type DekCache struct {
	kms                *kms.Client
	encryptingKmsKeyID string
	encryptingDek      *EncryptingDek
	encryptingDekMutex sync.RWMutex
	decryptingDeks     *lru.Cache[string, cipher.AEAD]
}

// NewDekCache creates a new DEK cache
func NewDekCache(kmsClient *kms.Client, encryptingKmsKeyID string) *DekCache {
	// cipher.AEAD:
	// 	16 byte interface
	// 	288 'gcm' object underneath
	// = 304 bytes
	// the lru cache itself will probably have some overhead as well, so call it 350.
	// 8192 entries is around 3 megabytes.
	decryptingDeks, err := lru.New[string, cipher.AEAD](8192)
	if err != nil {
		// the library only returns an error if size is <= 0
		panic(err)
	}
	return &DekCache{
		kms:                kmsClient,
		encryptingKmsKeyID: encryptingKmsKeyID,
		decryptingDeks:     decryptingDeks,
	}
}

// GetEncryptingDek returns the encrypting DEK, creating it if necessary
func (d *DekCache) GetEncryptingDek(ctx context.Context) (*EncryptingDek, error) {
	d.encryptingDekMutex.RLock()
	if d.encryptingDek != nil {
		defer d.encryptingDekMutex.RUnlock()
		return d.encryptingDek, nil
	}
	d.encryptingDekMutex.RUnlock()

	d.encryptingDekMutex.Lock()
	defer d.encryptingDekMutex.Unlock()

	// Double-check pattern
	if d.encryptingDek != nil {
		return d.encryptingDek, nil
	}

	start := time.Now()
	result, err := d.kms.GenerateDataKey(ctx, &kms.GenerateDataKeyInput{
		KeyId:   &d.encryptingKmsKeyID,
		KeySpec: types.DataKeySpecAes256,
	})
	if err != nil {
		log.Printf("Failed to create encrypting dek against KMS key '%s': %v", d.encryptingKmsKeyID, err)
		return nil, fmt.Errorf("failed to generate data key: %w", err)
	}

	if result.Plaintext == nil || result.CiphertextBlob == nil {
		return nil, fmt.Errorf("missing plaintext or ciphertext from KMS response")
	}

	// Create AES-GCM cipher for encryption
	block, err := aes.NewCipher(result.Plaintext)
	if err != nil {
		return nil, fmt.Errorf("failed to create AES cipher: %w", err)
	}

	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("failed to create GCM mode: %w", err)
	}

	d.encryptingDek = &EncryptingDek{
		Key:          aead,
		EncryptedDek: result.CiphertextBlob,
	}

	// Cache the decrypting version as well
	d.decryptingDeks.Add(string(result.CiphertextBlob), aead)

	log.Printf("Created encrypting dek against KMS key '%s' in %v", d.encryptingKmsKeyID, time.Since(start))
	return d.encryptingDek, nil
}

// GetDecryptingDek returns a decrypting DEK for the given encrypted DEK
func (d *DekCache) GetDecryptingDek(ctx context.Context, encryptedDek []byte) (cipher.AEAD, error) {
	key := string(encryptedDek)

	if cached, exists := d.decryptingDeks.Get(key); exists {
		return cached, nil
	}

	start := time.Now()
	result, err := d.kms.Decrypt(ctx, &kms.DecryptInput{
		CiphertextBlob: encryptedDek,
	})
	if err != nil {
		log.Printf("Failed to load decrypting dek: %v", err)
		return nil, fmt.Errorf("failed to decrypt DEK: %w", err)
	}

	if result.Plaintext == nil {
		return nil, fmt.Errorf("missing plaintext from KMS decrypt response")
	}

	block, err := aes.NewCipher(result.Plaintext)
	if err != nil {
		return nil, fmt.Errorf("failed to create AES cipher: %w", err)
	}

	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("failed to create GCM mode: %w", err)
	}

	d.decryptingDeks.Add(key, aead)

	log.Printf("Loaded decrypting dek in %v", time.Since(start))
	return aead, nil
}
